src/realm/sync/client.cpp

```diff
@@ -1779,6 +1807,10 @@ void SessionWrapper::report_progress(bool only_if_new_uploadable_data)
     if (!m_progress_handler)
         return;
 
+    // Ignore progress messages from before we first receive a DOWNLOAD message
+    if (!m_reliable_download_progress)
```

Comment:
this was just a refactoring moving this exact check from object-store callback. I've separated changes under separate commit here, so you may see that it doesn't change the observed behaviour on object-store side. There was also fixme in this function about this.

Reply:
how would you tell if there is progress on upload then?

Reply:
there are two scenarios where we can report upload progress: when commiting (this decreases the progress estimate) and when receiving download messages (this increases the progress estimate)

Reply:
it could be I don't fully get the question, but we can report upload progress without needing any download message (keep reporting the committed data)

Reply:
uploaded is changed when we receive download messages. The use case I'm thinking is that the user should still get progress messages even if they are offline (if they make changes of course)

Reply:
we could also say we only report progress when data is uploaded (irrespective to what is committed)

Reply:
we do both already, no? uploadable is updated and propagates to object store api on commit notification, and uploaded is updated when acknowledged by the server with DOWNLOAD message.

This is a long discussion already, i'm not sure what is being asked here. Probably, even more confusing the original name of this variable how it was: m_reliable_download_progress 

It's not exactly 'reliable' for anything, essentially it was just blocking progress reporting until session receives first ever DOWNLOAD message. Probably, to make sure what the state of sync progress from the server side?

Reply:
> essentially it was just blocking progress reporting until session receives first ever DOWNLOAD message.

This contradicts with progress being reported if the user is offline.

Reply:
yes, right, what i meant is that we invoke progress update because of uploadable changes on commit, but not until once again first download message is received. So this variable m_reliable_download_progress was just like initial guard to prevent any notifications. And once again i'm still not sure what was the idea behind this, i've just done refactoring moving it away from object store into sync impl, since tests in both places seemed to diverge. So what do you think essentially? Should we just remove it?

Reply:
ok.. i'm now very confused with this. I'll do what you ask, no problem, i'm not going to argue since looks like my understanding is a bit off here. That will require adapt the code once again in tests mostly. 

just one last time to educate me on this one, i seek to understand the sync impl a bit better with this:

```
// so the code was in sync impl before:
void SessionWrapper::report_progress(bool only_if_new_uploadable_data) {
    ...

    uint64_t progress_version = (m_reliable_download_progress ? 1 : 0);
    m_progress_handler(..., progress_version, ...);
}
// .. and in object-store: 
void SyncProgressNotifier::update(..., uint64_t download_version, ...) {
    // Ignore progress messages from before we first receive a DOWNLOAD message
    if (download_version == 0)
        return;

    ...
}
```

what this subtly breaking behaviour be if it is checked now at the beginning here:
```
void SessionWrapper::report_progress(bool only_if_new_uploadable_data) {
    if (!m_reliable_download_progress)
        return;
    ...
}
```

Reply:
Let's talk about this tomorrow.

Reply:
Talked offline and this change makes sense to me now.